LinuxAdapter.cs

```diff
@@ -0,0 +1,88 @@
+﻿using System;
+using System.DirectoryServices.Protocols;
+using System.Linq;
+using System.Net;
+using System.Security.Claims;
+using System.Security.Principal;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Authentication.Negotiate
+{
+    public static class LinuxAdapter
+    {
+        // example: machineAccount = "user@DOMAIN.com" machinePassword = "***"
+        public static Task OnAuthenticated(AuthenticatedContext context, string machineAccount, string machinePassword, bool resolveNestedGroups = true)
+        {
+            var domain = machineAccount.Substring(machineAccount.IndexOf('@') + 1);
+            var distinguishedName = domain.Split('.').Select(name => $"dc={name}").Aggregate((a, b) => $"{a},{b}");
+            var user = context.Principal.Identity.Name;
+            var userAccountName = user.Substring(0, user.IndexOf('@'));
+
+            LdapDirectoryIdentifier di = new LdapDirectoryIdentifier(domain, true, false);
+            NetworkCredential credential = new NetworkCredential(machineAccount, machinePassword);
+            using LdapConnection connection = new LdapConnection(di, credential);
+            connection.SessionOptions.ProtocolVersion = 3; //Setting LDAP Protocol to latest version
+            connection.Bind(); // This line actually makes the connection.
+            connection.Timeout = TimeSpan.FromMinutes(3);
+            string filter = $"(&(objectClass=user)(sAMAccountName={userAccountName}))"; // This is using ldap search query language, it is looking on the server for someUser
+            SearchRequest searchRequest = new SearchRequest(distinguishedName, filter, SearchScope.Subtree, null);
+            SearchResponse searchResponse = (SearchResponse)connection.SendRequest(searchRequest);
+
+            if (searchResponse.Entries.Count > 0)
+            {
+                var userFound = searchResponse.Entries[0]; //Get the object that was found on ldap
```

Comment:
Would there ever be more than one?

Reply:
I haven't run into any cases where this is true. But this would be a good question for @joperezr

Reply:
Depends on the search. If we're looking up by upn, which we get from the auth, then probably not. However, we should perhaps log if there's more than 1, to indicate weirdness. After well, we never expected guid duplication and that happened last year.

Reply:
basically what @blowdart said. LDAP doesn't allow two users with the same samaccountname, but it is possible that if you don't perform your search to be exact (like using contains or partial strings) you may get more than one result back. You ideally want to be the search to be exact so that you only get one user back, as well as log errors if that is not the case as suggested.

Reply:
Yup I added some logging for then this occurs, maybe I should bump it up to a warning or error though